File: 2025/04 April/07 Directed Graph Cycle/main.py
# Link: https://www.geeksforgeeks.org/problems/detect-cycle-in-a-directed-graph/1


class Solution:
    def isCycle(self, V, edges):
        in_degree = [0] * V
        graph = [[] for _ in range(V)]

        for u, v in edges:
            graph[u].append(v)
        for node_list in graph:
            for node in node_list:
                in_degree[node] += 1

        queue = []
        for i in range(V):
            if in_degree[i] == 0:
                queue.append(i)
        sorted_nodes = []
        while queue:
            current = queue.pop(0)
            sorted_nodes.append(current)
            for adj in graph[current]:
                in_degree[adj] -= 1
                if in_degree[adj] == 0:
                    queue.append(adj)
        return len(sorted_nodes) < V


if __name__ == "__main__":
    obj = Solution()
    V = 4
    edges = [[0, 1], [1, 2], [2, 3], [3, 3]]
    result = obj.isCycle(V, edges)
    if result:
        print("Cycle detected")
    else:
        print("No cycle detected")

    V = 3
    edges = [[0, 1], [1, 2]]
    result = obj.isCycle(V, edges)
    if result:
        print("Cycle detected")
    else:
        print("No cycle detected")

File: 2025/04 April/07 Directed Graph Cycle/main.cpp
// Link: https://www.geeksforgeeks.org/problems/detect-cycle-in-a-directed-graph/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    bool dfsCheck(int node, vector<int> adj[], vector<int> &vis, vector<int> &pathVis)
    {
        vis[node] = 1;
        pathVis[node] = 1;
        for (auto it : adj[node])
        {
            if (!vis[it])
            {
                if (dfsCheck(it, adj, vis, pathVis))
                {
                    return true;
                }
            }
            else if (pathVis[it])
            {
                return true;
            }
        }
        pathVis[node] = 0;
        return false;
    }

    bool isCyclic(int V, vector<vector<int>> &edges)
    {
        vector<int> adj[V];
        for (auto edge : edges)
        {
            adj[edge[0]].push_back(edge[1]);
        }
        vector<int> vis(V, 0);
        vector<int> pathVis(V, 0);
        for (int i = 0; i < V; i++)
        {
            if (!vis[i])
            {
                if (dfsCheck(i, adj, vis, pathVis))
                {
                    return true;
                }
            }
        }
        return false;
    }
};

int main()
{
    Solution s;
    int V = 4;
    vector<vector<int>> edges = {{0, 1}, {1, 2}, {2, 3}, {3, 3}};
    bool result = s.isCyclic(V, edges);
    cout << (result ? "Cycle detected" : "No cycle detected") << endl;

    V = 3;
    edges = {{0, 1}, {1, 2}};
    result = s.isCyclic(V, edges);
    cout << (result ? "Cycle detected" : "No cycle detected") << endl;

    return 0;
}